alf/optimizers/utils.py

```diff
@@ -31,3 +36,113 @@ def get_opt_arg(p: nn.Parameter, argname: str, default: Any = None):
+            update_rate: the update rate for computing moving averages of the
+                quantities needed by GNS. Generally, a smaller value (slower update)
+                makes the estimated GNS more biased (because quantities at different
+                training steps are averaged) while a larger value (quicker
+                update) makes it have more variances.
+            gradient_norm_clip: a clipping value for gradient norm outliers. If
+                None, no clipping is performed.
+        """
+        super().__init__()
+        self._name = name
+        assert 0 < batch_size_ratio < 0.5
+        self._batch_size_ratio = batch_size_ratio
+        self._grad_norm_clip = gradient_norm_clip
+        self._gradient_norm_averager = ScalarEMAverager(
+            update_rate=update_rate)
+        self._var_trace_averager = ScalarEMAverager(update_rate=update_rate)
+
+    def _calculate_gradient_norm(self, loss: torch.Tensor,
+                                 tensors: alf.nest.NestedTensor):
+        grads = alf.nest.utils.grad(tensors, loss.mean(), retain_graph=True)
+        norm2 = global_norm(grads)
+        if self._grad_norm_clip is not None:
+            norm2 = torch.clamp(norm2, max=self._grad_norm_clip)
+        return norm2**2
+
+    def forward(self, loss: torch.Tensor, tensors: alf.nest.NestedTensor):
+        """Given a loss tensor and a nest of tensors, return the estimated GNS.
+
+        Args:
+            loss: a loss tensor *before* taking the mean. Each entry of the tensor
+                represents an individual loss on a single training sample. Ideally,
+                the samples used for computing these losses should be sampled *with*
+                replacement independently. The loss can have a shape of either
+                ``[T,B]`` or ``[B]``. In the first case, we will only take
+                ``[0,B]`` for computing the gns due to sample correlation along
+                the temporal axis. On-policy algorithms might have noisy estimates
+                if the ``num_parallel_environments`` is small. Similarly, this
+                happens for a small batch size for off-policy algorithms.
+            tensors: a nest of tensors whose gradients are considered
+
+        Returns:
+            gns: the estimated gradient noise scale (a scalar). A smaller value
+                means more effective grad steps.
+        """
+        assert loss.ndim in [1, 2], "loss must be a rank-1 or -2 tensor!"
+        if loss.ndim == 2:
+            loss = loss[0, :]
+        B = loss.shape[0]
+        shuffled_loss = loss[torch.randperm(B)]
+
+        b = max(1, int(B * self._batch_size_ratio))
+        B -= b
+
+        B_norm2 = self._calculate_gradient_norm(shuffled_loss[b:], tensors)
+        b_norm2 = self._calculate_gradient_norm(shuffled_loss[:b], tensors)
+
+        gradient_norm = (B * B_norm2 - b * b_norm2) / (B - b)
+        var_trace = (b_norm2 - B_norm2) / (1. / b - 1. / B)
```

Comment:
Cause this is quite different from the original formula, do you have some derivation showing that this is what is needed?

Reply:
> Cause this is quite different from the original formula, do you have some derivation showing that this is what is needed?

This is the simplified GNS defined in Appendix A. Compared to the complex one, it doesn't require computing the hessian matrix.